File: test/samples/floats/floats.js
"use strict";

/**
 * @param {number} f a float
 * @param {number} g a float, greater than f
 * @param {{x: number, y: number, z: number}} point some coordinates
 * @param {number} n a number
 * @param {Array.<number>} floatList a list of floats
 * @param {Array.<number>} otherList a list of floats
 * @param {Array.<{integer: number, char: string, float: number}>} inlined some inlined structs
 * @param {{'integer 2': number, string: string, 'float 2': number}} multiline a multiline struct
 * @returns {void}
 */
function floats(f, g, point, n, floatList, otherList, inlined, multiline) {
    /* TODO Parsing is often easy, reprint mode is harder */
}

function main(stdin) {
    let line = 0;

    const f = Number(stdin[line++]);
    const g = Number(stdin[line++]);
    const words = stdin[line++].split(" ");
    const point = {
        x: Number(words[0]),
        y: Number(words[1]),
        z: Number(words[2])
    };
    const n = Number(stdin[line++]);
    const floatList = stdin[line++].split(" ", n).map(Number);
    const otherList = stdin[line++].split(" ", 9).map(Number);
    const inlined = [];
    for (let i = 0; i < 3; i++) {
        const words1 = stdin[line++].split(" ");
        const j = {
            integer: Number(words1[0]),
            char_: words1[1],
            float_: Number(words1[2])
        };
        inlined.push(j);
    }
    const multiline = {};
    multiline.integer2 = Number(stdin[line++]);
    multiline.string = stdin[line++];
    multiline.float2 = Number(stdin[line++]);
    floats(f, g, point, n, floatList, otherList, inlined, multiline);
}

let stdin = "";
process.stdin.on("data", data => stdin += data.toString())
             .on("end", () => main(stdin.split("\n")));

File: test/samples/floats/floats.hs
import Control.Monad (replicateM)

-- | Represents coordinates
data Coordinates = Coordinates
  { x :: Double  -- ^ X
  , y :: Double  -- ^ Y
  , z :: Double  -- ^ Z
  }

-- | Mix of fields that go on one line
data InlinedMix = InlinedMix
  { integer :: Int     -- ^ an integer
  , char    :: Char    -- ^ a char
  , float   :: Double  -- ^ a float
  }

-- | a struct of chars
data MultilineMix = MultilineMix
  { integer2 :: Int     -- ^ an other integer
  , string   :: String  -- ^ a string of size 5
  , float2   :: Double  -- ^ an other float
  }

floats :: Double        -- ^ a float
       -> Double        -- ^ a float, greater than f
       -> Coordinates   -- ^ some coordinates
       -> Int           -- ^ a number
       -> [Double]      -- ^ a list of floats
       -> [Double]      -- ^ a list of floats
       -> [InlinedMix]  -- ^ some inlined structs
       -> MultilineMix  -- ^ a multiline struct
       -> String        -- ^ TODO
-- Parsing is often easy, reprint mode is harder
floats f g point n floatList otherList inlined multiline = "TODO"

main :: IO ()
main = do
  f <- fmap read getLine
  g <- fmap read getLine
  point <- readCoordinates
  n <- fmap read getLine
  floatList <- fmap (map read . words) getLine
  otherList <- fmap (map read . words) getLine
  inlined <- replicateM 3 readInlinedMix
  multiline <- readMultilineMix
  putStrLn $ floats f g point n floatList otherList inlined multiline
  where
    readCoordinates = fmap ((\[a, b, c] -> Coordinates (read a) (read b) (read c)) . words) getLine
    readInlinedMix = fmap ((\[a, b, c] -> InlinedMix (read a) (head b) (read c)) . words) getLine
    readMultilineMix = MultilineMix <$> fmap read getLine <*> getLine <*> fmap read getLine
